Remove Transport dropdown in computer setup, assume SSH

danielhollas opened the pull request:
It is not clear whether AiiDA Computer setup ever worked with any transport other than SSH,
but in any case the widget is clearly designed with SSH in mind so unless a clear need is found we can simply remove the 'Transport type" dropdown. Note the the value of this widget has not been used anywhere in the code, which is telling.

The localhost computer is setup in the AiiDAlab Docker image by default, and in the unlikely case that another one needs to be setup, such user probably knows how to set it up via verdi command line interface.

Fixes #417

yakutovicha reviewed (commented):
I think this is good for now.

However, later I am introducing the [`ssh_light`](https://github.com/aiidateam/aiida-core/pull/6083) which will revert things to `Dropdown`. Then, there will be `firecrest`. It would take some more work to adopt them, though.

Therefore, for now, it is good to keep it as `HTML` to signify that widget is not dealing with other cases.

I would only request to keep the `self.transport.value = "core.ssh"` line and set the original value to `core.ssh` too.

It was issued later on. Also, settings downloaded from the [code database](https://github.com/aiidateam/aiida-code-registry) override the value. Herein, `SSH` as the value may cause trouble.

danielhollas commented:
> However, later I am introducing the https://github.com/aiidateam/aiida-core/pull/6083 which will revert things to Dropdown. Then, there will be firecrest. It would take some more work to adopt them, though.

Ah, that's good to know. I also found out that the transport value is indeed passed to the computer setup function, I did not notice that before since it is accessed via `getattrs`

https://github.com/aiidalab/aiidalab-widgets-base/blob/413c26c76ec438b9c70afe3437420923f738e50b/aiidalab_widgets_base/computational_resources.py#L898

> I would only request to keep the self.transport.value = "core.ssh" line and set the original value to core.ssh too.
> It was issued later on. Also, settings downloaded from the [code database](https://github.com/aiidateam/aiida-code-registry) override the value. Herein, SSH as the value may cause trouble.

Another good point. As a side note, I am not a fan of this design, `core.ssh` is not very user friendly, we're essentially exposing the user to internal AiiDA details. But that's for another time...

Given all of the above, I've reverted the original change and instead fix the underlying bug and made the `core.local` transport work. I also added a unit test. @yakutovicha could you please test locally?

danielhollas commented on aiidalab_widgets_base/computational_resources.py:
```diff
@@ -858,6 +858,17 @@ def _configure_computer(self, computer: orm.Computer):
+            "safe_interval": self.safe_interval.value,
+        }
+        user = orm.User.collection.get_default()
+        computer.configure(user=user, **authparams)
```
Note that we need to catch any potential errors here. Since this was not done in the original function, I am keeping that for separate PR.

danielhollas commented on aiidalab_widgets_base/computational_resources.py:
```diff
@@ -902,7 +913,14 @@ def on_setup_computer(self, _=None):
+                self._configure_computer_local(computer)
+            else:
+                self.message = (
+                    f"WARNING: Transport value {self.transport.value} not recognized!"
```
I think it would be safer to throw an error here. I guess a computer that is not configured is not usable?

yakutovicha replied:
> I guess a computer that is not configured is not usable?

One can configure it later, of course. But that is not the purpose of the widget. The widget should create a computer that is ready to be used.

> I think it would be safer to throw an error here.

I agree. I made a suggestion below.

danielhollas replied:
Agreed, donw

yakutovicha commented on aiidalab_widgets_base/computational_resources.py:
```diff
@@ -902,7 +913,14 @@ def on_setup_computer(self, _=None):
+            if self.transport.value == "core.local":
+                self._configure_computer_local(computer)
+            else:
+                self.message = (
+                    f"WARNING: Transport value {self.transport.value} not recognized!"
+                )
```
```suggestion
                raise common.ValidationError(f"transport value {self.transport.value} not recognized!")
```

yakutovicha reviewed (commented):
Thanks, @danielhollas. As you've suggested, let's not allow creating a computer if its configuration is not working. The rest is good, I tested the widget locally and it works as expected 👍 .

danielhollas commented on aiidalab_widgets_base/computational_resources.py:
```diff
@@ -902,16 +936,15 @@ def on_setup_computer(self, _=None):
             computer = computer_builder.new()
-            self._configure_computer(computer)
+            self._configure_computer(computer, self.transport.value)
+            computer.store()
```
I've moved this here since it can also fail.

Strangely, when a user provides empty workdir, the store command raises ValidationError, but the computer is still stored. That to me seems like `aiida-core` bug.

danielhollas commented on aiidalab_widgets_base/computational_resources.py:
```diff
@@ -902,16 +936,15 @@ def on_setup_computer(self, _=None):
             RuntimeError,
         ) as err:
-            self.message = f"Failed to setup computer {type(err).__name__}: {err}"
+            self.message = f"Computer setup failed! {type(err).__name__}: {err}"
```
Made this shorter, since long error messages do not fit the widget. We'll need a better solution for this.

danielhollas commented on aiidalab_widgets_base/computational_resources.py:
```diff
@@ -937,7 +970,7 @@ def _reset(self):
         self.hostname.value = ""
         self.description.value = ""
-        self.work_dir.value = ""
+        self.work_dir.value = "/home/{username}/aiida_run"
```
work_dir cannot be empty. Should we provide some default value like this or not??

CC @yakutovicha @unkcpz

danielhollas replied:
I decided to revert this change. I think this needs to be a conscious and informed decision from the  user. On some HPCs, setting this to home directory could potentially lead to excessive network communication.

danielhollas reviewed (commented):
@yakutovicha I've made some more improvements, can you take a look?

yakutovicha commented on aiidalab_widgets_base/computational_resources.py:
```diff
@@ -817,8 +817,17 @@ def observe_memory_per_machine(change):
+            self._configure_computer_local(computer)
+        else:
+            msg = f"invalid transport type '{transport}'"
+            raise common.exceptions.ValidationError(msg)
```
`ValidationError` is a part of the `common`'s API:
```suggestion
            raise common.ValidationError(msg)
```

yakutovicha commented on aiidalab_widgets_base/computational_resources.py:
```diff
@@ -858,6 +866,17 @@ def _configure_computer(self, computer: orm.Computer):
+            "use_login_shell": self.use_login_shell.value,
+            "safe_interval": self.safe_interval.value,
+        }
+        user = orm.User.collection.get_default()
```
Maybe do that inside `_configure_computer` method and then pass the `user` argument to `_configure_computer_local` and `_configure_computer_ssh`? We will then use it also for the other transport types.

yakutovicha reviewed (changes requested):
Thanks, @danielhollas. I suggested a few minor changes. I also tested it, all good.

yakutovicha reviewed (approved):
I tested the latest changes - all good 👍. Ship it 🚢 !

danielhollas commented:
Great, thanks @yakutovicha!
